treq/testing.py

```diff
@@ -205,3 +207,212 @@ def __init__(self, resource):
+    """
+    def __repr__(self):
+        return "ANYTHING"
+
+    def __eq__(self, _):
+        return True
+
+    def __ne__(self, _):
+        return False
+
+ANY = _any()
+
+
+class SequenceStringStubs(object):
+    """
+    Takes a sequence of::
+
+        [((method, url, params, headers, data), (code, headers, body)),
+         ...]
+
+    Expects the requests to arrive in sequence order.  If there are no more
+    responses, or the request's paramters do not match the next item's expected
+    request paramters, raises :obj:`AssertionError`.
+
+    :ivar list sequence: The sequence of expected request arguments mapped to
+        stubbed responses
+    :ivar testcase: A :obj:`TestCase` that can be used to report failures in
+        matching requests against expected requests.  Raising exceptions does
+        not work, because :obj:`Resource.render` will just convert that into
+        a 500 response.
+    """
+    def __init__(self, sequence, testcase):
+        self._sequence = sequence
+        self._testcase = testcase
+
+    def consumed(self):
+        """
+        :return: `bool` representing whether the entire sequence has been
+            consumed.  This is useful in tests to assert that the expected
+            requests have all been made.
+        """
+        return len(self._sequence) == 0
+
+    @contextmanager
+    def consume(self):
+        """
+        Usage::
+
+            stub_treq = StubTreq(StringStubbingResource(
+                sequence_stubs.get_response_for))
+            with sequence_stubs.consume():
+                stub_treq.get('http://fakeurl.com')
+                stub_treq.get('http://another-fake-url.com')
+
+        If there are still remaining expected requests to be made in the
+        sequence, fails the provided test case.
+
+        :return: a context manager that can be used to ensure all expected
+            requests have been made.
+        """
+        yield
+        if not self.consumed():
+            self._testcase.fail("\n".join(
+                ["Not all expected requests were made.  Still expecting:"] +
+                ["- {0}(url={1}, params={2}, headers={3}, data={4})".format(
+                    *expected) for expected, _ in self._sequence]))
+
+    def get_response_for(self, method, url, params, headers, data):
+        """
+        :return: the next response in the sequence, provided that the
+            parameters match the next in the sequence.
+        """
+        if len(self._sequence) == 0:
+            self._testcase.addCleanup(
+                self._testcase.fail,
+                "No more requests expected, but request {0!r} made.".format(
+                    (method, url, params, headers, data)))
+            return (500, {}, "StubbingError")
+
+        expected, response = self._sequence[0]
+        e_method, e_url, e_params, e_headers, e_data = expected
+
+        checks = [
+            (e_method is ANY or e_method.lower() == method.lower(), "method"),
```

Comment:
Just to clarify, you mean just for the method right?  The comment above also mentioned about the headers, and while I agree with you that the ANY checks everywhere is weird, I'd also like it to be consistent.

It seems necessary for params, because there needs to be some kind of way to distinguish between no params and any params (speaking of which do you think I should also accept `None`?), and no data and any data.  And its' easier to document that `ANY` can be used for all these values, instead of "the first parameter is required, the second and third parameters can be ANY, but for the fourth parameter `{}` represents `ANY`, and the fifth parameter can be `ANY`.

Maybe we can switch the order of the expected parameters so that method and headers are clustered together, so the special casing is all the first two parameters?  I was hoping to take them in the same order as the treq `get/put/post/etc.` paramters though.

What do you think?